components/brave_wallet/common/bitcoin_utils_unittest.cc

```diff
@@ -0,0 +1,145 @@
+
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/common/bitcoin_utils.h"
+
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace brave_wallet {
+
+TEST(BitcoinUtilsUnitTest, Bip0173TestVectors1) {
+  // https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#examples
+
+  std::vector<uint8_t> pubkey;
+  ASSERT_TRUE(base::HexStringToBytes(
+      "0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798",
+      &pubkey));
+
+  EXPECT_EQ(PubkeyToSegwitAddress(pubkey, true),
+            "tb1qw508d6qejxtdg4y5r3zarvary0c5xw7kxpjzsx");
+  EXPECT_EQ(PubkeyToSegwitAddress(pubkey, false),
+            "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4");
+
+  // TODO(apaymyshev): support P2WSH.
+}
+
+TEST(BitcoinUtilsUnitTest, Bip0173TestVectors2) {
+  // https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#test-vectors
+
+  struct ValidTestCase {
+    const char* address;
+    const char* opcodes;  // not relevant parts of scriptpubkey.
+    const char* pubkey_hash_hex;
+  } test_cases[] = {
+      {
+          "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4",
+          "0014",
+          "751e76e8199196d454941c45d1b3a323f1433bd6",
+      },
+      {
+          "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3q0sl5k7",
+          "0020",
+          "1863143c14c5166804bd19203356da136c985678cd4d27a1b8c6329604903262",
+      },
+      {
+          "bc1pw508d6qejxtdg4y5r3zarvary0c5xw7kw508d6qejxtdg4y5r3zarvary0c5xw7k"
+          "7grplx",
+          "5128",
+          "751e76e8199196d454941c45d1b3a323f1433bd6751e76e8199196d454941c45d1b3"
+          "a323f1433bd6",
+      },
+      {
+          "BC1SW50QA3JX3S",
+          "6002",
+          "751e",
+      },
+      {
+          "bc1zw508d6qejxtdg4y5r3zarvaryvg6kdaj",
+          "5210",
+          "751e76e8199196d454941c45d1b3a323",
+      },
+      {
+          "tb1qqqqqp399et2xygdj5xreqhjjvcmzhxw4aywxecjdzew6hylgvsesrxh6hy",
+          "0020",
+          "000000c4a5cad46221b2a187905e5266362b99d5e91c6ce24d165dab93e86433",
+      }};
+
+  for (auto& test_case : test_cases) {
+    SCOPED_TRACE(test_case.address);
+    bool testnet = base::StartsWith(test_case.address, "tb",
+                                    base::CompareCase::INSENSITIVE_ASCII);
+    std::vector<uint8_t> pubkey_hash;
+    ASSERT_TRUE(
+        base::HexStringToBytes(test_case.pubkey_hash_hex, &pubkey_hash));
+    auto decoded_address = DecodeBitcoinAddress(test_case.address, testnet);
+    ASSERT_TRUE(decoded_address);
+    EXPECT_EQ(decoded_address->pubkey_hash, pubkey_hash);
```

Comment:
`address_type` should be checked.

Reply:
https://github.com/brave/brave-core/pull/17978/commits/2f117c534c8236292c2bbc209de5195830ccb9c5